scripts/symbol_check/check_unsupported_symbol.sh

```diff
@@ -0,0 +1,210 @@
+        echo "Skipping symbol check: Failed to find ENTRYPOINT to given Docker image $image_name" >&2
+        exit 1
+    fi
+
+    executable=($args)
+    executable=${executable[0]:1:-1}
+
+    # check if path absolute
+    if [[ ! "$executable" =~ ^/.* ]]; then
+        echo "Skipping symbol check: ENTRYPOINT must be absolute. Instead got: $executable" >&2
+        exit 1
+    fi
+
+    echo $executable
+}
+
+# Given a executable/shared object
+# Return a list of UNDEFINED symbols that it uses
+get_list_symbols() {
+    local executable=$1
+
+    # List dynamic external Undefined symbols, filter only undefined symbols
+    raw_symbols=$(nm -D -g -u $executable 2>/dev/null | grep " U "); rc=${PIPESTATUS[0]}
+    if [[ $rc -ne 0 ]]; then
+        # Skip this shared object
+        echo "get_list_symbols: Failed to read symbols from $executable" >&2
+        return 1
+    fi
+
+    parsed_symbols=""
+    for line in ${raw_symbols}; do
+        if [[ "U" != $line ]]; then
+            parsed_symbols+="$(parse_symbol $line) "
+        fi
+    done
+
+    echo "${parsed_symbols}"
+}
+
+# Given an executable, return a list of shared objects used by it and its dependencies
+get_list_shared_objects() {
+    local appdir=$1
+    # Note: The path of executable should be relative to appdir
+    local executable=$2
+
+    # This list stores all shared objects used by the executable and its dependencies
+    # All paths should be relative to appdir
+    list_shared_objects="$executable"
+
+    # A queue used for BFS
+    queue="$executable"
+
+    # Current search depth
+    level=0
+    while [[ ! -z "$queue" && $level -lt $SEARCH_DEPTH ]]; do
+        # For each shared object in the queue, list its dependencies
+        # And add them to the queue
+        queue_next_level=""
+        for object in $queue; do
+            # Use ldd to list all linked .so, grab lines contain "=>" but not "not found"
+            output="$(ldd $appdir$object 2>/dev/null | grep "=>" | grep -i -v "not" | cut -d' ' -f 3)"; rc=${PIPESTATUS[0]}
+            if [[ $rc -eq 0 ]]; then
+                queue_next_level+="$output "
+            fi
+        done
+
+        list_shared_objects=$(echo "$list_shared_objects $queue_next_level" | tr ' ' '\n' | sort | uniq)
+        queue=$queue_next_level
+
+        # increase level
+        level=$((level+1))
+    done
+
+    echo "${list_shared_objects}"
+}
+
+# Given an executable
+# Return a list of symbols used by the executable and its dependencies
+get_list_symbols_of_executable() {
+    local appdir=$1
+    local executable=$2
+    # 1. find all lib dependencies using ldd
+    list_so=$(get_list_shared_objects "$appdir" "$executable")
+
+    # 2. check symbols of each dependency using nm
+    list_symbols=""
+
+    for each in $list_so; do
+        tmp_symbols="$(get_list_symbols $appdir$each) "; rc=${PIPESTATUS[0]}
+        if [[ $rc -eq 0 ]]; then
+            list_symbols+="$tmp_symbols "
+        fi
+    done
+
```

Comment:
It's probably more performant to remove the duplicates before the next step `filter_symbols`.